tools/base/abstract.py

```diff
@@ -0,0 +1,163 @@
+from abc import ABCMeta
+from typing import Any, Dict, List, Tuple, Type, Union
+
+
+class Implementer(type):
+    """Metaclass for implementers of an Abstract interface
+
+    Any `Abstraction` classes that are listed in `__implements__` for the
+    class are added as bases (ie added to the class's inheritance).
+
+    Any docs for methods are copied from the interface method to the
+    implementation if the implementation method has no docs of its own.
+
+    For example:
+
+    ```
+
+    from tools.base.abstract import Abstraction, Implementer
+
+
+    class AFoo(metaclass=Abstraction):
+
+        @abstractmethod
+        def do_something(self):
+            \"""Do something\"""
+            raise NotImplementedError
+
+    class Foo(metaclass=Implementer)
```

Comment:
Should this just be documented with decorator given this is how it will mostly be used?

Reply:
this is demonstrating what the decorator is doing under the hood - ie the addition of the `__implements__` and metaclass

it defo doesnt want the decorator to be demonstrated here (that is demoed on the deco function) - perhaps we can remove it if its in some way confusing - it was added  before i added the decorator